File: src/server/src/models/MaterialCategory.ts
import mongoose from "mongoose";
import { ModelsNames } from "./_modelsNames";
export interface MaterialCategoryAttrs {
  name: string;
}
export interface MaterialCategoryDoc
  extends mongoose.Document,
    MaterialCategoryAttrs {
  createdAt: Date;
  updatedAt: Date;
}

export const MaterialCategorySchema = new mongoose.Schema<MaterialCategoryDoc>(
  {
    name: {
      type: String,
      required: true,
      unique: true,
    },
  },
  {
    timestamps: true,
    toJSON: {
      transform(_, ret) {
        ret.id = ret._id;
        delete ret._id;
        delete ret.__v;
      },
    },
  }
);
export interface MaterialCategoryModel
  extends mongoose.Model<MaterialCategoryDoc> {
  build(attrs: MaterialCategoryAttrs): MaterialCategoryDoc;
}

MaterialCategorySchema.statics.build = (attrs: MaterialCategoryAttrs) => {
  return new MaterialCategory(attrs);
};
export const MaterialCategory = mongoose.model<
  MaterialCategoryDoc,
  MaterialCategoryModel
>(ModelsNames.MaterialCategory, MaterialCategorySchema);

File: src/mobile_app/lib/src/api/models/MaterialCategory.dart
class MaterialCategory {
  String _id;
  String name;

  MaterialCategory(
    this._id,
    this.name,
  );
  get id => _id;

  MaterialCategory.fromJson(Map<String, dynamic> json)
      : this(
          json['_id'],
          json['name'],
        );

  Map<String, dynamic> toJson() => {
        'id': _id,
        'name': name,
      };

  @override
  String toString() {
    return 'MaterialCategory{id: $_id, name: $name}';
  }
}
